modules/porous_flow/src/materials/PorousFlowMassFraction.C

```diff
@@ -0,0 +1,124 @@
+    _mass_frac_old(declarePropertyOld<std::vector<std::vector<Real> > >("PorousFlow_mass_frac")),
+    _grad_mass_frac(declareProperty<std::vector<std::vector<RealGradient> > >("PorousFlow_grad_mass_frac")),
+    _dmass_frac_dvar(declareProperty<std::vector<std::vector<std::vector<Real> > > >("dPorousFlow_mass_frac_dvar")),
+
+    _yaqi_hacky(false),
+
+    _num_passed_mf_vars(coupledComponents("mass_fraction_vars"))
+{
+  if (_num_phases < 1 || _num_components < 1)
+    mooseError("PorousFlowMassFraction: The Dictator proclaims that the number of phases is " << _num_phases << " and the number of components is " << _num_components << ", and stipulates that you should not use PorousFlowMassFraction in this case");
+  if (_num_passed_mf_vars != _num_phases*(_num_components - 1))
+    mooseError("PorousFlowMassFraction: The number of mass_fraction_vars is " << _num_passed_mf_vars << " which must be equal to the Dictator's num_phases (" << _num_phases << ") multiplied by num_components-1 (" << _num_components - 1 << ")");
+
+  _mf_vars_num.resize(_num_passed_mf_vars);
+  _mf_vars.resize(_num_passed_mf_vars);
+  _grad_mf_vars.resize(_num_passed_mf_vars);
+  for (unsigned i = 0; i < _num_passed_mf_vars; ++i)
+  {
+    _mf_vars_num[i] = coupled("mass_fraction_vars", i);
+    _mf_vars[i] = &coupledNodalValue("mass_fraction_vars", i);
+    _grad_mf_vars[i] = &coupledGradient("mass_fraction_vars", i);
+  }
+}
+
+void
+PorousFlowMassFraction::initQpStatefulProperties()
+{
+  const unsigned int num_var = _dictator_UO.numVariables();
+  _mass_frac[_qp].resize(_num_phases);
+  _mass_frac_old[_qp].resize(_num_phases);
+  _grad_mass_frac[_qp].resize(_num_phases);
+  _dmass_frac_dvar[_qp].resize(_num_phases);
+  for (unsigned int ph = 0; ph < _num_phases; ++ph)
+  {
+    _mass_frac[_qp][ph].resize(_num_components);
+    _mass_frac_old[_qp][ph].resize(_num_components);
+    _grad_mass_frac[_qp][ph].resize(_num_components);
+    _dmass_frac_dvar[_qp][ph].resize(_num_components);
+    for (unsigned int comp = 0; comp < _num_components; ++comp)
+      _dmass_frac_dvar[_qp][ph][comp].assign(num_var, 0.0);
+  }
+
+  // the derivative matrix is fixed for all time
+  // so it can be built here instead of in computeQpProperties
+  unsigned int i = 0;
+  for (unsigned int ph = 0; ph < _num_phases; ++ph)
+  {
+    for (unsigned int comp = 0; comp < _num_components - 1; ++comp)
+    {
+      if (_dictator_UO.isPorousFlowVariable(_mf_vars_num[i]))
+      {
+        // _mf_vars[i] is a PorousFlow variable
+        const unsigned int pf_var_num = _dictator_UO.porousFlowVariableNum(_mf_vars_num[i]);
+        _dmass_frac_dvar[_qp][ph][comp][pf_var_num] = 1.0;
+        _dmass_frac_dvar[_qp][ph][_num_components - 1][pf_var_num] = -1.0;
+      }
+      i++;
+    }
+  }
+  if (!_yaqi_hacky)
+    build_mass_frac(_qp);
+}
+
+void
+PorousFlowMassFraction::computeQpProperties()
+{
+  build_mass_frac(_qp);
+
+  if (_yaqi_hacky)
```

Comment:
OK - remind me what this is for. I know it has to do with resolving dependencies but I'd like to fix this if possible. In fact our `old` properties really should be read only. What can I do in the framework to make your life a little easier?